tools/make_c_api.lua

```diff
@@ -0,0 +1,432 @@
+    for field in msg:fields() do
+      table.insert(fields_number_order, field)
+    end
+    table.sort(fields_number_order, function(a, b)
+      return a:number() < b:number()
+    end)
+
+    append('struct %s {\n', msgname)
+    for _, field in ipairs(fields_layout_order) do
+      field_count = field_count + 1
+
+      if field:type() == upb.TYPE_MESSAGE then
+        submsg_count = submsg_count + 1
+        submsg_set[field:subdef()] = true
+      end
+
+      if has_hasbit(field) then
+        hasbit_indexes[field] = hasbit_count
+        hasbit_count = hasbit_count + 1
+      end
+
+      append('  %s %s;\n', ctype(field), field:name())
+    end
+    append('};\n\n')
+
+    if submsg_count > 0 then
+      -- TODO(haberman): could save a little bit of space by only generating a
+      -- "submsgs" array for every strongly-connected component.
+      local submsgs_array_name = msgname .. "_submsgs"
+      submsgs_array_ref = "&" .. submsgs_array_name .. "[0]"
+      append('static const upb_msglayout_msginit_v1 *const %s[%s] = {\n',
+             submsgs_array_name, submsg_count)
+
+      -- Create a deterministically-sorted array of submessage entries.
+      local submsg_array = {}
+      for k, v in pairs(submsg_set) do
+        table.insert(submsg_array, k)
+      end
+      table.sort(submsg_array, function(a, b)
+        return a:full_name() < b:full_name()
+      end)
+
+      for i, submsg in ipairs(submsg_array) do
+        append('  &%s_msginit,\n', to_cident(submsg:full_name()))
+        submsg_indexes[submsg] = i - 1
+      end
+
+      append('};\n\n')
+    end
+
+    if field_count > 0 then
+      local fields_array_name = msgname .. "__fields"
+      fields_array_ref = "&" .. fields_array_name .. "[0]"
+      append('static const upb_msglayout_fieldinit_v1 %s[%s] = {\n',
+             fields_array_name, field_count)
+      for _, field in ipairs(fields_number_order) do
+        local submsg_index = "-1"
+        local oneof_index = "UPB_NOT_IN_ONEOF"
+        if field:type() == upb.TYPE_MESSAGE then
+          submsg_index = submsg_indexes[field:subdef()]
+        end
+        -- TODO(haberman): oneofs.
+        append('  {%s, offsetof(%s, %s), %s, %s, %s, %s, %s},\n',
+               field:number(),
+               msgname,
+               field:name(),
+               hasbit_indexes[field] or "-1",
+               oneof_index,
+               submsg_index,
+               field:descriptor_type(),
+               field:label())
+      end
+      append('};\n\n')
+    end
+
+    append('const upb_msglayout_msginit_v1 %s_msginit = {\n', msgname)
+    append('  %s,\n', submsgs_array_ref)
+    append('  %s,\n', fields_array_ref)
+    append('  NULL, /* TODO. oneofs */\n')
+    append('  NULL, /* TODO. default_msg */\n')
+    append('  UPB_ALIGNED_SIZEOF(%s), %s, %s, %s, %s\n',
+           msgname, field_count,
+           0, -- TODO: oneof_count
+           'false', -- TODO: extendable
+           'true' -- TODO: is_proto2
```

Comment:
The main problem for the TODO is that Lua doesn't currently have access to this information. A follow-up CL that is in progress fixes this.